pfsimulator/parflow_lib/reservoir_package.c

```diff
@@ -0,0 +1,533 @@
+// Temp solution while I get a more permanent var dz fix checked in
+double GetSubgridVolume(Subgrid *subgrid, ProblemData* problem_data){
+  double dx = SubgridDX(subgrid);
+  double dy = SubgridDY(subgrid);
+  double dz = SubgridDZ(subgrid);
+  GrGeomSolid *gr_domain = problem_data->gr_domain;
+
+  double volume = 0;
+  Grid           *grid = VectorGrid(problem_data->rsz);
+  SubgridArray   *subgrids = problem_data->dz_mult->grid->subgrids;
+  Subgrid        *tmp_subgrid;
+  Subvector      *dz_sub;
+  Subvector      *rsz_sub;
+  int subgrid_index;
+  int index_space_z;
+  bool found_index_space_z = false;
+  ForSubgridI(subgrid_index, subgrids) {
+    tmp_subgrid = SubgridArraySubgrid(subgrids, subgrid_index);
+    Subvector *dz_mult_subvector = VectorSubvector(problem_data->dz_mult, subgrid_index);
+    double* dz_mult_data = SubvectorData(dz_mult_subvector);
+    Subgrid *intersection = IntersectSubgrids(subgrid, tmp_subgrid);
+    int nx = SubgridNX(intersection);
+    int ny = SubgridNY(intersection);
+    int nz = SubgridNZ(intersection);
+    int r = SubgridRZ(intersection);
+    int ix = SubgridIX(intersection);
+    int iy = SubgridIY(intersection);
+    int iz = SubgridIZ(intersection);
+    int i, j, k;
+    GrGeomInLoop(i, j, k, gr_domain, r, ix, iy, iz, nx, ny, nz,
+                 {
+
+                   int index = SubvectorEltIndex(dz_mult_subvector, i, j, k);
+                   //real space z will give us the cell center...not the cell bottom. So we correct for that
+                   //I think if we call this function generically this line can seg fault. Need to make sure
+                   // index is good for generic case. I have ensured we only call in cases this can not seg
+                   // fault currently
+                   double dz_mult = dz_mult_data[index];
+                   volume += dz_mult * dx * dy * dz;
+                 });
+  }
+  return volume;
+
+};
+
+
+
+typedef struct {
+    int num_phases;
+    int num_contaminants;
+
+    /* reservoir info */
+    int num_units;
+    int num_reservoirs;
+
+    int       *type;
+    void     **data;
+
+    /* Timing Cycle information */
+    int num_cycles;
+
+    int       *interval_divisions;
+    int      **intervals;
+    int       *repeat_counts;
+
+    NameArray reservoir_names;
+} PublicXtra;
+
+typedef void InstanceXtra;
+
+typedef struct {
+    char    *name;
+    double intake_x_location, intake_y_location;
+    int has_secondary_intake_cell;
+    double secondary_intake_x_location, secondary_intake_y_location;
+    double release_x_location, release_y_location;
+    double max_storage, min_release_storage, storage, release_rate;
+} Type0;                      /* basic vertical reservoir */
+
+/*--------------------------------------------------------------------------
+ * ReservoirPackage
+ *--------------------------------------------------------------------------*/
+
+bool subgrid_lives_on_this_rank(Subgrid* subgrid, Grid *grid){
```

Comment:
New code should have doxygen comments for methods.

https://github.com/parflow/parflow/blob/master/CONTRIBUTING.md#source-code-documentation